hilti/src/rt/tests/library.cc

```diff
@@ -0,0 +1,225 @@
+
+    auto* s = ::dlsym(RTLD_DEFAULT, "foo");
+
+    if ( auto* error = ::dlerror() ) {
+        return result::Error(error);
+    }
+
+    return s;
+};
+
+// RAII helper to set an environment variable.
+class Env {
+public:
+    Env(std::string k, const std::string_view& v) {
+        if ( auto prev = ::getenv(k.data()) )
+            _prev = {k, prev};
+        else
+            _prev = {k, std::nullopt};
+
+        REQUIRE_EQ(::setenv(k.c_str(), v.data(), 1), 0);
+    }
+
+    ~Env() {
+        const auto& [k, v] = _prev;
+
+        if ( v )
+            REQUIRE_EQ(::setenv(k.c_str(), v->c_str(), 1), 0);
+        else
+            REQUIRE_EQ(::unsetenv(k.c_str()), 0);
+    }
+
+private:
+    std::pair<std::string, std::optional<std::string>> _prev;
+};
+
+// RAII helper to create a temporary directory.
+class TemporaryDirectory {
+public:
+    TemporaryDirectory() {
+        const auto tmpdir = std::filesystem::temp_directory_path();
+        auto template_ = (tmpdir / "hilti-rt-test-XXXXXX").native();
+        auto path = ::mkdtemp(template_.data());
+        REQUIRE_NE(path, nullptr);
+        _path = path;
+    }
+
+    ~TemporaryDirectory() {
+        if ( ! std::filesystem::exists(_path) )
+            return;
+
+        // Make sure we have permissions to remove the directory.
+        std::filesystem::permissions(_path, std::filesystem::perms::all);
+        for ( const auto& entry : std::filesystem::recursive_directory_iterator(_path) )
+            std::filesystem::permissions(entry, std::filesystem::perms::all);
+
+        std::filesystem::remove_all(_path);
+    }
+
+    const auto& path() const { return _path; }
+
+private:
+    std::filesystem::path _path;
+};
+
+// RAII helper to redirect output.
+class CaptureIO {
+public:
+    CaptureIO(std::ostream& stream) : _old(stream.rdbuf(_buffer.rdbuf())), _stream(&stream) {}
+    ~CaptureIO() { _stream->rdbuf(_old); }
+
+    auto str() const { return _buffer.str(); }
+
+private:
+    std::stringstream _buffer = std::stringstream{};
+    std::streambuf* _old = nullptr;
+    std::ostream* _stream = nullptr;
+};
+
+TEST_SUITE_BEGIN("Library");
+
+TEST_CASE("construct") {
+    Library _(dummy); // Does not throw
+    CHECK_THROWS_WITH_AS(Library("/does/not/exist"), "no such library: \"/does/not/exist\"", const EnvironmentError&);
+
+    SUBCASE("TMPDIR does not exist") {
+        Env _("TMPDIR", "/does/not/exist");
+        try {
+            Library _(dummy);
+            FAIL("expected exception not thrown");
+        } catch ( const EnvironmentError& e ) {
+            CAPTURE(e.description());
+            CHECK(startsWith(e.description(),
+                             fmt("could not add library \"%s\": could not create temporary file: ", dummy.c_str())));
+        } catch ( ... ) {
+            FAIL("unexpected exception thrown");
+        }
+    }
+
+    SUBCASE("cannot write TMPDIR") {
+        // NOTE: This test likely does not work if run as `root` as `root`
```

Comment:
want to add uid==0 test and just skip in that case? Tests shouldn't be running as root, but who knows ...

Reply:
Ideally we'd skip this test if we are in a privileged context where file permissions are ignored, but it is really hard to check for in a generic, platform-independent way so I left it out.

The correct way to check for this under Linux would be to check whether the calling process has the `CAP_DAC_OVERRIDE` capability (which `root` usually has); I am not sure what the equivalent concept of that would be on macos. That being said, I think the most likely scenario where the process has such privileges would be when running in Docker containers without explicitly configured `USER` (like we now do for our CI images). I don't want to worry about other scenarios as I'd like to claim that e.g., a user running these unit tests as `root` on macos is clearly "holding it wrong".

I'd prefer we don't try to do more here and instead have anybody running into issues stumble into the comment I added.